File: retrocpu_boot_monitor/mn1613/test/handshake/handshake_lcd1_test.ts
/**
 * g_bios_lcd_control（CPU→IO コマンド 17h）
 * 根拠: HandShake.mdc「LCD制御」/ boot_monitor.mdc / test_framework.mdc
 */
import {
  createSessionFromSettings,
  expect,
  test,
  type IoBoardHandshakeMock,
  type Mn1613AsmSession,
} from "../../../../retrocpu_test_framework/src/index.js";
import {
  mn1613MonHandshakeSettings,
  withMn1613CpuLog,
} from "../mn1613_mon_settings.js";

const BASE_REGS = {
  R3: 0x3333,
  R4: 0x4444,
} as const;

const session: Mn1613AsmSession = createSessionFromSettings(
  withMn1613CpuLog(mn1613MonHandshakeSettings, import.meta.url),
);

/**
 * g_main 済み＋ ioMock handshake で 1 ケースを実行する。
 * @param fn 本体（session / mock 利用可）
 */
async function withCase(
  fn: (s: Mn1613AsmSession, mock: IoBoardHandshakeMock) => Promise<void>,
): Promise<void> {
  session.reload();
  try {
    await session.runInit();
    await fn(session, session.requireHandshakeMock());
  } finally {
    await session.detachIoMock();
  }
}

/**
 * g_bios_lcd_control を呼び、CPU→IO 1 トランザクションと並行する。
 * @param mock IO モック
 * @param kind 種別（R0）
 * @param argA 引数A（R1）
 * @param argB 引数B（R2 Bit8–9）
 * @param argC 引数C（R2 Bit0–7）
 */
async function callLcd1(
  mock: IoBoardHandshakeMock,
  kind: number,
  argA: number,
  argB: number,
  argC: number,
): Promise<void> {
  await Promise.all([
    session.call("g_bios_lcd_control", {
      registers: {
        ...BASE_REGS,
        R0: kind,
        R1: argA,
        R2: ((argB & 3) << 8) | (argC & 0xff),
      },
    }),
    mock.handleOneRequest(),
  ]);
}

test("17h フレーム（kind/argA/argB/argC）を正しく送る", async () => {
  await withCase(async (s, mock) => {
    await callLcd1(mock, 3, 0x05, 0x01, 0x0f);

    const last = mock.state.log.at(-1);
    expect(last).toBeTruthy();
    expect(last!.dir).toBe("cpu_to_io");
    expect(last!.cmd).toBe(0x17);
    expect(Array.from(last!.frame)).toEqual([0x17, 0x03, 0x05, 0x01, 0x0f]);

    const status = last!.response?.[0] ?? 0;
    s.expectRegisters({ R0: status & 0xff });
  });
});

test("R3/R4 は呼び出しの前後で保たれる", async () => {
  await withCase(async (s, mock) => {
    await callLcd1(mock, 2, 0x07, 0x00, 0x00);
    s.expectRegisters({ R3: BASE_REGS.R3, R4: BASE_REGS.R4 });
  });
});

File: retrocpu_boot_monitor/test/mn1613/handshake/handshake_lcd1_test.rs
use std::sync::Arc;

use retrocpu_test_framework_rs::{
    create_session_from_settings, CallOptions, CallRegisters, CodeTestIoMockEntry, FrameworkError,
    IoBoardHandshakeMock, JsonTestSettings, Mn1613AsmSession,
};

fn handshake_settings() -> JsonTestSettings {
    let mut s = super::mn1613_rs_settings();
    s.io_mock = Some(vec![CodeTestIoMockEntry::Handshake]);
    s
}

fn with_case<F>(f: F) -> Result<(), FrameworkError>
where
    F: FnOnce(&mut Mn1613AsmSession, Arc<IoBoardHandshakeMock>) -> Result<(), FrameworkError>,
{
    let mut s = create_session_from_settings(&handshake_settings(), None)?;
    s.run_init()?;
    let mock = s.require_handshake_mock()?;
    let ret = f(&mut s, Arc::clone(&mock));
    s.detach_io_mock();
    ret
}

fn lcd_control_regs(kind: u16, arg_a: u16, arg_b: u16, arg_c: u16) -> CallRegisters {
    CallRegisters {
        r0: Some(kind),
        r1: Some(arg_a),
        r2: Some(((arg_b & 0x03) << 8) | (arg_c & 0xff)),
        r3: Some(0x3333),
        r4: Some(0x4444),
        ..Default::default()
    }
}

fn call_lcd1(
    s: &mut Mn1613AsmSession,
    mock: &Arc<IoBoardHandshakeMock>,
    kind: u16,
    arg_a: u16,
    arg_b: u16,
    arg_c: u16,
) -> Result<(), FrameworkError> {
    mock.run_with_cpu_to_io_request(|| {
        let _ = s.call(
            "g_bios_lcd_control",
            CallOptions {
                registers: Some(lcd_control_regs(kind, arg_a, arg_b, arg_c)),
                ..Default::default()
            },
        )?;
        Ok(())
    })?;
    Ok(())
}

#[test]
fn lcd_control_sends_kind_arga_argb_argc_frame() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        call_lcd1(s, &mock, 3, 0x05, 0x01, 0x0f)?;
        s.expect_registers(
            &CallRegisters {
                r0: Some(0),
                ..Default::default()
            },
            None,
        )
    })
}

#[test]
fn lcd_control_preserves_r3_r4() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        call_lcd1(s, &mock, 2, 0x07, 0x00, 0x00)?;
        s.expect_registers(
            &CallRegisters {
                r0: Some(0),
                r3: Some(0x3333),
                r4: Some(0x4444),
                ..Default::default()
            },
            None,
        )
    })
}
